libecho/src/EchoProtocol.cc

```diff
@@ -0,0 +1,220 @@
+#include "EchoProtocol.h"
+#include <cassert>
+#include <QDebug>
+
+using namespace std::chrono_literals;
+
+static constexpr size_t PACKET_SIZE = 50;
+
+// NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init)
+EchoProtocol::EchoProtocol(int winsize, int send_freq, int recv_freq, int lim) {
+    big_win_size = static_cast<double>(winsize) / 44100 * (PACKET_SIZE + 10) * 12s;
+    buffer = new uint8_t[2 * PACKET_SIZE];
+    closed = false;
+    connection = std::unique_ptr<EchoRawConnection>(EchoRawConnection::getBitEchoRawConnection(
+        winsize, send_freq, recv_freq, lim));
+    thr[0] = thr[1] = nullptr;
+    is_connected = false;
+    status = 0;
+    lastPacketAcked = 0;
+    number = 0;
+}
+
+EchoProtocol::~EchoProtocol() {
+    delete[] buffer;
+}
+
+void EchoProtocol::listen() {
+    qDebug() << "listen";
+    is_connected = false;
+    thr[0] = new std::thread{ &EchoProtocol::receivingThread, this, true };
+    while (!is_connected) {}
+    thr[1] = new std::thread{ &EchoProtocol::sendingThread, this, false };
+    qDebug() << "(listen) connected successfully";
+}
+
+void EchoProtocol::connect() {
+    qDebug() << "connect";
+    thr[1] = new std::thread{ &EchoProtocol::sendingThread, this, true };
+    std::this_thread::sleep_for(big_win_size);
+    thr[0] = new std::thread{ &EchoProtocol::receivingThread, this, false };
+    qDebug() << "connected successfully";
+}
+
+void EchoProtocol::close() {
+    qDebug() << "close";
+    status = 4;
+    closed = true;
+    if (thr[0] != nullptr) {
+        thr[0]->join();
+        delete thr[0];
+    }
+    if (thr[1] != nullptr) {
+        thr[1]->join();
+        delete thr[1];
+    }
```

Comment:
To powinieneś móc przywrócić do poprzedniej formy - deletowanie NULLi jest bezpieczne.